topi/include/topi/nn/dilate.h

```diff
@@ -0,0 +1,86 @@
+/*!
+ *  Copyright (c) 2017 by Contributors
+ * \brief Dilate op constructions
+ * \file nn/dilate.h
+ */
+#ifndef TOPI_NN_DILATE_H_
+#define TOPI_NN_DILATE_H_
+
+#include <string>
+#include <vector>
+
+#include "tvm/tvm.h"
+#include "tvm/ir_pass.h"
+#include "topi/tags.h"
+
+namespace topi {
+namespace nn {
+using namespace tvm;
+
+/*!
+* \brief Create a new experssion of the intersection of all
+* conditions in the arguments.
+*
+* \return The intersection expression
+*/
+Expr all(Array<Expr> args) {
+  CHECK_GT(args.size(), 0) << "all requires at least one argument";
+
+  Expr ret = args[0];
+  for (size_t i = 1; i < args.size(); ++i) {
+    ret = ret && args[i];
+  }
+  return ret;
+}
+
+/*!
+* \brief Dilate data with zeros
+*
+* \param x The input tensor, this can have any number of
+* dimensions and any layout.
+* \param strides Dilation stride for each dimension. Stride 1
+* means no dilation.
+* \param name The name of the operation
+* \param tag The tag to mark the operation
+*
+* \return The output tensor.
+*/
+inline Tensor dilate(const Tensor& x,
+                     std::vector<int> strides,
```

Comment:
Change most of the std::vector<int> arguments to Array<Expr>

Reply:
How should an array of ints be represented? Should it be an Array<Expr> filled with constants? (Array<int> fails to compile)

Reply:
re: /reduction.h (I can't reply directly for some reason), I've pulled out most of the duplication into 2 common functions: MakeReduceAxes and MakeReduceTargetShape they should be in the latest commit

Reply:
Array of <Expr> note that strides can be passed in as symbolic expressions, although not frequently used

Reply:
Sorry array of what? Github may have broken your comment.

I see that strides could be expressions, but nn/dilate.h:69 relies on being able to evaluate strides[i]. This and surrounding lines are:
```
        if (strides[i] != 1) {
          index_tuple.push_back(indices[i] / strides[i]);
          not_zero.push_back((indices[i] % strides[i]) == 0);
        } else {
          index_tuple.push_back(indices[i]);
        }
```

Would a correct implementation of this with changing the signature to `Array<Expr> strides` be this?
```
        if (IsConstInt(strides[i]) && GetConstInt(strides[i]) != 1) {
          index_tuple.push_back(indices[i] / strides[i]);
          not_zero.push_back((indices[i] % strides[i]) == 0);
        } else {
          index_tuple.push_back(indices[i]);
        }
```

Reply:
yes, let us do that